crates/storage/nippy-jar/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+            }
+
+            row_number += 1;
+            if row_number == total_rows {
+                break
+            }
+
+            column_iterators = iterators.into_iter();
+        }
+
+        // Write offset index to file
+        bincode::serialize_into(File::create(self.index_path())?, &values_offsets)?;
+
+        Ok(())
+    }
+
+    /// Safety checks before creating and returning a [`File`] handle to write data to.
+    fn freeze_check(
+        &mut self,
+        columns: &Vec<impl IntoIterator<Item = Vec<u8>>>,
+    ) -> Result<File, NippyJarError> {
+        if columns.len() != self.columns {
+            return Err(NippyJarError::ColumnLenMismatch(self.columns, columns.len()))
+        }
+
+        if let Some(compression) = &self.compressor {
+            if !compression.is_ready() {
+                return Err(NippyJarError::CompressorNotReady)
+            }
+        }
+
+        Ok(File::create(self.data_path())?)
+    }
+
+    /// Writes all necessary configuration to file.
+    fn freeze_config(&self, handle: &mut File) -> Result<(), NippyJarError> {
+        // TODO Split Dictionaries and Bloomfilters Configuration so we dont have to load everything
+        // at once
+        Ok(bincode::serialize_into(handle, &self)?)
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum NippyJarError {
+    #[error("err")]
+    Err,
+    #[error("err")]
+    Disconnect(#[from] std::io::Error),
+    #[error("err")]
+    Bincode(#[from] Box<bincode::ErrorKind>),
+    #[error("Compression was enabled, but it's not ready yet.")]
+    CompressorNotReady,
+    #[error("Decompression was enabled, but it's not ready yet.")]
+    DecompressorNotReady,
+    #[error("Number of columns does not match. {0} != {1}")]
+    ColumnLenMismatch(usize, usize),
+    #[error("UnexpectedMissingValue row: {0} col:{1}")]
+    UnexpectedMissingValue(u64, u64),
+}
+
+pub struct NippyJarCursor<'a> {
+    config: &'a NippyJar,
+    zstd_decompressors: Option<Mutex<Vec<Decompressor<'a>>>>,
+    data_offsets: Vec<u64>,
+    data_handle: File,
+    row: u64,
+    col: u64,
+}
+
+impl<'a> std::fmt::Debug for NippyJarCursor<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "NippyJarCursor {{ config: {:?} }}", self.config)
+    }
+}
+
+impl<'a> NippyJarCursor<'a> {
+    pub fn new(
+        config: &'a NippyJar,
+        zstd_decompressors: Option<Mutex<Vec<Decompressor<'a>>>>,
+    ) -> Result<Self, NippyJarError> {
+        let data_offsets = bincode::deserialize_from(File::open(config.index_path())?)?;
+
+        Ok(NippyJarCursor {
+            config,
+            zstd_decompressors,
+            data_offsets,
+            data_handle: File::open(config.data_path())?,
+            row: 0,
+            col: 0,
+        })
+    }
+
+    pub fn reset(&mut self) {
+        self.row = 0;
+        self.col = 0;
+    }
+
+    pub fn next_row(&mut self) -> Result<Option<Vec<Vec<u8>>>, NippyJarError> {
```

Comment:
a lot of fine handling in these parts...let's talk about it when it's in a more complete state to make sure everyone on the team has a solid understanding

Reply:
yeah, this is the most basic query to allow it for tests.

 it will probably not be actually used, since we want finger detail: `want a value from a column, and not the whole row`